### ISSUE TITLE
Keras with TF on a multi GPU machine

### ISSUE THREAD

**Issue by somya12** [keras-team/keras#9856]
I am training a [mnist](https://github.com/keras-team/keras/blob/master/examples/mnist_mlp.py) model using keras (2.1.5) with tensorflow (1.6) as backend. I am using g3.8xlarge which has two Tesla M60 GPUs with around 7613 MB of GPU memory. When I start training the model, I can see that both the GPUs are getting used. Below is the output of nvidia-smi command -

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.81                 Driver Version: 384.81                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla M60           On   | 00000000:00:1D.0 Off |                    0 |
| N/A   62C    P0    86W / 150W |   7353MiB /  7613MiB |     74%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla M60           On   | 00000000:00:1E.0 Off |                    0 |
| N/A   49C    P0    79W / 150W |   7353MiB /  7613MiB |     67%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     10772      C   ...a-4.2.0-py-3.5.3-dl-gpu-full/bin/python  7342MiB |
|    1     10772      C   ...a-4.2.0-py-3.5.3-dl-gpu-full/bin/python  7342MiB |
+-----------------------------------------------------------------------------+
```

I have a couple of questions -

1. On a multi GPU node, how is the data distributed among the GPUs?

2. I haven't written any code for distributed training (specifying cluster spec etc.), still my training is distributed since it's happening on 2 GPUs. Is it using the parameter server architecture to compute final gradients?

3. If it's not using parameter server architecture, how are the two GPU communicating with each other?

4. I have also seen cases where during the entire model training, the volatile GPU utilisation stays at 0% for one GPU and remains as high as 84% for the other one. The memory utilization is almost same for both the GPUs. Are there cases when only one GPU is used to train and the other one just keeps the data in memory?

This might not be the right forum to ask this but any documentation or examples would be really helpful. Thanks!

**Comment by dennis-ec** [comment]
Tensorflow always tries to allocate all gpu memory of every available machine. You can limit the visibile device by setting the env variable CUDA_VISIBLE_DEVICES. Also, it's a good idea to use allow growth in your tensorflow session(keras will relato to it automatically)
 ```
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
 sess = tf.Session(config=config)
```
This way will use only as much gpu memory as needed.

What I don't understand is why your second GPU is working. I only know the case, where tensorflow occupy every gpu's memory but only calculate on the first one
